### PR TITLE
feat: add support for the temporal evolution of relationships #367

### PR DESCRIPTION
**Pull request by bobeal**

### CODE REVIEW

File: search-service/src/main/kotlin/com/egm/stellio/search/service/TemporalEntityAttributeService.kt [hunk 1]
```diff
@@ -72,50 +83,46 @@ class TemporalEntityAttributeService(
...
+            ) {
+                it.getAttributeInstances().map { instance ->
+                    Pair(it.name, toAttributeMetadata(instance))
                 }
-            )
+            }.filter {
+                it.second.isValid
+            }.map {
+                Pair(it.first, it.second.toEither().orNull()!!)
+            }
```
Comment (HoucemKacem) [inline search-service/src/main/kotlin/com/egm/stellio/search/service/TemporalEntityAttributeService.kt:97]:
```
.ifEmpty {
    return Mono.just(0)
}
```

File: search-service/src/main/kotlin/com/egm/stellio/search/service/EntityEventListenerService.kt [hunk 2]
```diff
@@ -162,51 +177,83 @@ class EntityEventListenerService(
...
+    internal fun toAttributeMedata(jsonNode: JsonNode): Validated<String, AttributeMetadata> {
+        if (!jsonNode.has("observedAt")) {
+            return "Ignoring append event for $jsonNode, it has no observedAt information".invalid()
+        }
+        val attributeTypeAsText = jsonNode["type"].asText()
+        val attributeType = kotlin.runCatching {
+            TemporalEntityAttribute.AttributeType.valueOf(attributeTypeAsText)
+        }.getOrNull() ?: return "Unsupported attribute type: $attributeTypeAsText".invalid()
+        val attributeValue = when (attributeType) {
+            TemporalEntityAttribute.AttributeType.Relationship -> Pair(jsonNode["object"].asText(), null)
```
[the remark this answers is not in the document]
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/service/EntityEventListenerService.kt:236]: Hmm, the point is that the payload comes from the entity service, so it has already been validated and we can be sure a relationship has an `object` attribute. Here I just want to validate we can and want handle it (it has an `observedAt`, it has a type we support, ...).

File: search-service/src/main/kotlin/com/egm/stellio/search/service/EntityEventListenerService.kt [hunk 3]
```diff
@@ -112,47 +130,44 @@ class EntityEventListenerService(
...
-                    )
-            )
-            attributeInstanceService.create(attributeInstance)
-                .then(
-                    temporalEntityAttributeService.updateEntityPayload(
-                        entityId,
-                        serializeObject(compactedJsonLdEntity)
+        when (val extractedAttributeMetadata = toAttributeMedata(attributeValuesNode)) {
+            is Invalid -> {
+                logger.info(extractedAttributeMetadata.e)
```
[the remark this answers is not in the document]
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/service/EntityEventListenerService.kt:135]: Here, `Invalid` is just invalid with respect to the temporal behavior. For instance, if it has no `observedAt`, it is considered as Invalid but this is not a problem (it is just not a temporal attribute).

File: search-service/src/main/kotlin/com/egm/stellio/search/service/EntityEventListenerService.kt [hunk 4]
```diff
@@ -162,51 +177,83 @@ class EntityEventListenerService(
...
-        val rawAttributeValue = attributeValuesNode["value"]
-        val parsedAttributeValue =
-            if (rawAttributeValue.isNumber)
-                Pair(null, valueToDoubleOrNull(rawAttributeValue.asDouble()))
-            else
-                Pair(valueToStringOrNull(rawAttributeValue.asText()), null)
-        val datasetId = attributeValuesNode["datasetId"]?.asText()?.toUri()
+    }
 
+    internal fun toAttributeMedata(jsonNode: JsonNode): Validated<String, AttributeMetadata> {
```
[the remark this answers is not in the document]
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/service/EntityEventListenerService.kt:227]: Indeed.

File: search-service/src/main/kotlin/com/egm/stellio/search/service/TemporalEntityAttributeService.kt (hunk 1 above)
Comment (bobeal) [reply search-service/src/main/kotlin/com/egm/stellio/search/service/TemporalEntityAttributeService.kt:97]: nice one!

**Review by HoucemKacem — APPROVED** [review]

**Pull request merged**